presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java

```diff
@@ -439,24 +439,42 @@ public void testTtlAwareScheduling()
                 queryManager,
                 new SimpleTtlNodeSelectorConfig().setFallbackToSimpleNodeSelection(true));
 
+        NodeScheduler fallbackDisabledNodeScheduler = new NodeScheduler(
+                new LegacyNetworkTopology(),
+                nodeManager,
+                new NodeSelectionStats(),
+                nodeSchedulerConfig,
+                nodeTaskMap,
+                nodeTtlFetcherManager,
+                queryManager,
+                new SimpleTtlNodeSelectorConfig().setFallbackToSimpleNodeSelection(false));
+
         // Query is estimated to take 20 mins and has been executing for 3 mins, i.e, 17 mins left
         // So only node2 and node3 have enough TTL to run additional work
         Session session = sessionWithTtlAwareSchedulingStrategyAndEstimatedExecutionTime(new Duration(20, TimeUnit.MINUTES));
-        NodeSelector nodeSelector = nodeScheduler.createNodeSelector(session, CONNECTOR_ID);
+        NodeSelector nodeSelector = fallbackEnabledNodeScheduler.createNodeSelector(session, CONNECTOR_ID);
         queryManager.setExecutionTime(new Duration(3, TimeUnit.MINUTES));
         assertEquals(ImmutableSet.copyOf(nodeSelector.selectRandomNodes(3)), ImmutableSet.of(node2, node3));
 
         // Query is estimated to take 5 hours and has been executing for 1 hour, i.e, 4 hours left
         // No nodes will have enough TTL, so we fall back to simple node selection as per the config above
         session = sessionWithTtlAwareSchedulingStrategyAndEstimatedExecutionTime(new Duration(5, TimeUnit.HOURS));
-        nodeSelector = nodeScheduler.createNodeSelector(session, CONNECTOR_ID);
+        nodeSelector = fallbackEnabledNodeScheduler.createNodeSelector(session, CONNECTOR_ID);
         queryManager.setExecutionTime(new Duration(1, TimeUnit.HOURS));
         assertEquals(ImmutableSet.copyOf(nodeSelector.selectRandomNodes(3)), ImmutableSet.of(node1, node2, node3));
 
+        // Query is estimated to take 20 mins and has been executing for 3 mins, i.e, 17 minutes left
+        // So only node2 and node3 have enough TTL to run additional work, however, they are also in the excluded nodes list
+        // So no nodes are selected
+        session = sessionWithTtlAwareSchedulingStrategyAndEstimatedExecutionTime(new Duration(5, TimeUnit.HOURS));
+        nodeSelector = fallbackDisabledNodeScheduler.createNodeSelector(session, CONNECTOR_ID);
+        queryManager.setExecutionTime(new Duration(1, TimeUnit.HOURS));
+        assertEquals(ImmutableSet.copyOf(nodeSelector.selectRandomNodes(3, ImmutableSet.of(node2, node3))), ImmutableSet.of());
```

Comment:
Maybe I am mistaken, but would it be simpler to say nodeSelector.selectRandomNodes(3, ImmutableSet.of(node2, node3)).isEmpty() ?

Reply:
Often assert equals on lists will tell you the contents, i.e. what specifically is missing or present, which may be helpful when looking at the test failure.